File: benchmarks/pandas/bench_numeric_ops_log2_exp.py
"""
Benchmark: np.log2, np.log10, np.exp, np.sign applied to pandas Series and DataFrame.

Mirrors tsb seriesLog2, seriesLog10, seriesExp, seriesSign and their DataFrame variants.
Uses 100k-row data to match the TypeScript benchmark.
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

# Positive values for log2/log10; any values for exp/sign
data = [(i + 1) * 0.1 for i in range(SIZE)]
s = pd.Series(data, dtype=float)
df = pd.DataFrame({
    "a": [(i + 1) * 0.1 for i in range(SIZE)],
    "b": [(i + 1) * 0.2 for i in range(SIZE)],
})

# Warm-up
for _ in range(WARMUP):
    np.log2(s)
    np.log10(s)
    np.exp(s)
    np.sign(s)
    np.log2(df)
    np.log10(df)
    np.exp(df)
    np.sign(df)

start = time.perf_counter()
for _ in range(ITERATIONS):
    np.log2(s)
    np.log10(s)
    np.exp(s)
    np.sign(s)
    np.log2(df)
    np.log10(df)
    np.exp(df)
    np.sign(df)
total_ms = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "numeric_ops_log2_exp",
    "mean_ms": total_ms / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total_ms,
}))

File: benchmarks/tsb/bench_numeric_ops_log2_exp.ts
/**
 * Benchmark: seriesLog2 / seriesLog10 / seriesExp / seriesSign and DataFrame variants.
 *
 * Mirrors numpy/pandas element-wise math functions on 100k-row data:
 *   - np.log2(s), np.log10(s), np.exp(s), np.sign(s)
 *   - DataFrame.apply(np.log2), etc.
 *
 * Outputs JSON: {"function": "numeric_ops_log2_exp", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import {
  Series,
  DataFrame,
  seriesLog2,
  seriesLog10,
  seriesExp,
  seriesSign,
  dataFrameLog2,
  dataFrameLog10,
  dataFrameExp,
  dataFrameSign,
} from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

// Positive values for log2/log10; any values for exp/sign
const data = Array.from({ length: SIZE }, (_, i) => (i + 1) * 0.1);
const s = new Series({ data });
const df = DataFrame.fromColumns({
  a: Array.from({ length: SIZE }, (_, i) => (i + 1) * 0.1),
  b: Array.from({ length: SIZE }, (_, i) => (i + 1) * 0.2),
});

for (let i = 0; i < WARMUP; i++) {
  seriesLog2(s);
  seriesLog10(s);
  seriesExp(s);
  seriesSign(s);
  dataFrameLog2(df);
  dataFrameLog10(df);
  dataFrameExp(df);
  dataFrameSign(df);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  seriesLog2(s);
  seriesLog10(s);
  seriesExp(s);
  seriesSign(s);
  dataFrameLog2(df);
  dataFrameLog10(df);
  dataFrameExp(df);
  dataFrameSign(df);
}
const total_ms = performance.now() - start;

console.log(
  JSON.stringify({
    function: "numeric_ops_log2_exp",
    mean_ms: total_ms / ITERATIONS,
    iterations: ITERATIONS,
    total_ms,
  }),
);
